### ISSUE TITLE
Miner crashes at (Invalid header on mined block) err="block in the future"

### ISSUE THREAD

**Issue by msolima** [ethereum/go-ethereum#13876]
#### System information

Geth version: 1.6.0-unstable
OS & Version: Linux
Commit hash : (if `develop`)
9aca9e6deb243b87cc75325be593a3b0c2f0a113

#### Expected behaviour
I have a private network with 2 geth instances connected to it
When I try to start a 3rd instance as a miner using the following:

`geth  --rpcport "8547" --datadir "./datadir/" --port "30305" --networkid 22 --mine --minerthreads=1 --etherbase=0xXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX --bootnodes "enode://XXXXXXX" `

I expect the miner to start mining, and continue to mine and process all transaction

#### Actual behaviour
Currently the miner starts mining 

```
INFO [04-06|13:46:56] Starting mining operation 
INFO [04-06|13:46:56] Commit new mining work                   number=1 txs=0 uncles=0 elapsed=118.478µs
INFO [04-06|13:46:56] Successfully sealed new block            number=1 hash=b1a4cd…efe37e
INFO [04-06|13:46:56] 🔨 mined potential block                  number=1 hash=b1a4cd…efe37e
INFO [04-06|13:46:56] Commit new mining work                   number=2 txs=0 uncles=0 elapsed=183.846µs
INFO [04-06|13:46:58] Successfully sealed new block            number=2 hash=a79a43…9bd710
INFO [04-06|13:46:58] 🔨 mined potential block                  number=2 hash=a79a43…9bd710
INFO [04-06|13:46:58] Commit new mining work                   number=3 txs=0 uncles=0 elapsed=144.137µs
INFO [04-06|13:46:59] Successfully sealed new block            number=3 hash=20fbe7…583ce0
```

then after a 1 or 2 mins it halts and stops processing/mining on my private network with the following error:

```
INFO [04-06|13:47:44] Commit new mining work                   number=37 txs=0 uncles=0 elapsed=109.409µs
INFO [04-06|13:47:44] Successfully sealed new block            number=37 hash=a9e8d1…64b7cc
ERROR[04-06|13:47:44] Invalid header on mined block            err="block in the future"
```

Am I missing something ? Please help

Same issue is happening when I try to do miner.start() from the console of a geth connected instance

**Comment by karalabe** [comment]
Ah yes, we also hit this today. A huge consensus refactor was merged into Geth yesterday I believe to cater for different consensus engines, and it broke the miner for low difficulty networks. There's a PR https://github.com/ethereum/go-ethereum/pull/13870 to fix it, but it still needs a review confirm to merge. Feel free to run that PR, it should be stable enough (given that you're running master branch anyway) :)

**Comment by karalabe** [comment]
This should be fixed on latest master. Please check and report if you see anything like this. I really appreciate errors around this refactor :)

**Comment by msolima** [comment]
@karalabe just verified the this! works great!
Thanks :+1:

**Comment by msolima** [comment]
@karalabe you mentioned that you guys made changes to the mining processes for low difficulty networks. 
I'm having an issue where my contracts on my private network never get mined.
and then I get `Error: Contract transaction couldn't be found after 50 blocks undefined`

I'm using 
Geth
Version: 1.6.0-unstable
Git Commit: cc13d576f07fb6803e09fb42880591a67b8b0ef6

my genesis block has:

```
"difficulty": "0x400",
"gasLimit": "0x8000000",
```

and this is how am deploying the contract


```
var greeter = SampleContract.new(_greeting,{from:web3.eth.accounts[0], data: code, gas: 4000000}, function(e, contract){
    if(!e) {

        if(!contract.address) {
            console.log("Contract transaction send: TransactionHash: " + contract.transactionHash + " waiting to be mined...");

        } else {
            console.log("Contract mined! Address: " + contract.address);
            console.log(contract);
        }

    }else{
        console.log(e)
    }
})
```

any idea what's happening ? are you guys seeing this issue?

**Comment by karalabe** [comment]
How many nodes are you running? What's the topology? How fast are blocks mined? What's the current gas limit?

**Comment by msolima** [comment]
I'm running 3 nodes and 1 bootnode.

here's an output of my miner that am running on one of the nodes:
```

INFO [04-11|17:17:28] 🔗 block reached canonical chain          number=5344 hash=ee4046…d97ca2
INFO [04-11|17:17:28] 🔨 mined potential block                  number=5349 hash=4d6299…cc3737
INFO [04-11|17:17:28] Commit new mining work                   number=5350 txs=0 uncles=0 elapsed=135.489µs
INFO [04-11|17:17:32] Successfully sealed new block            number=5350 hash=b64699…e149c9
INFO [04-11|17:17:32] 🔗 block reached canonical chain          number=5345 hash=927fda…1dd14b
INFO [04-11|17:17:32] 🔨 mined potential block                  number=5350 hash=b64699…e149c9
INFO [04-11|17:17:32] Commit new mining work                   number=5351 txs=0 uncles=0 elapsed=131.078µs
INFO [04-11|17:17:36] Successfully sealed new block            number=5351 hash=44a845…f68b9a
INFO [04-11|17:17:36] 🔗 block reached canonical chain          number=5346 hash=627e72…43d3ab
INFO [04-11|17:17:36] 🔨 mined potential block                  number=5351 hash=44a845…f68b9a
INFO [04-11|17:17:36] Commit new mining work                   number=5352 txs=0 uncles=0 elapsed=118.884µs
INFO [04-11|17:17:48] Successfully sealed new block            number=5352 hash=2b483d…6acb66
INFO [04-11|17:17:48] 🔗 block reached canonical chain          number=5347 hash=f1b8ab…dd851a
INFO [04-11|17:17:48] 🔨 mined potential block                  number=5352 hash=2b483d…6acb66
INFO [04-11|17:17:48] Commit new mining work                   number=5353 txs=0 uncles=0 elapsed=143.227µs
INFO [04-11|17:17:53] Successfully sealed new block            number=5353 hash=d6c7cc…c87da5
INFO [04-11|17:17:53] 🔗 block reached canonical chain          number=5348 hash=7eddc0…302c70
INFO [04-11|17:17:53] 🔨 mined potential block                  number=5353 hash=d6c7cc…c87da5
INFO [04-11|17:17:53] Commit new mining work                   number=5354 txs=0 uncles=0 elapsed=130.011µs
INFO [04-11|17:17:54] Successfully sealed new block            number=5354 hash=8161c5…533cc6
INFO [04-11|17:17:54] 🔗 block reached canonical chain          number=5349 hash=4d6299…cc3737
INFO [04-11|17:17:54] 🔨 mined potential block                  number=5354 hash=8161c5…533cc6
INFO [04-11|17:17:54] Commit new mining work                   number=5355 txs=0 uncles=0 elapsed=166.668µs
INFO [04-11|17:17:57] Successfully sealed new block            number=5355 hash=357efc…9b14d3
INFO [04-11|17:17:57] 🔗 block reached canonical chain          number=5350 hash=b64699…e149c9
INFO [04-11|17:17:57] 🔨 mined potential block                  number=5355 hash=357efc…9b14d3
INFO [04-11|17:17:57] Commit new mining work                   number=5356 txs=0 uncles=0 elapsed=156.429µs

```

Current gas limit is 4712388

**Comment by karalabe** [comment]
Sooo... could you try latest master (it's enough to update the miner)? :D This was a loooong standing bug with networks consisting of a single miner (the miner never considered itself synced, and rejected transactions). A fix for this was merged in 2 days ago I think.

**Comment by msolima** [comment]
ok I'll try that. I don't need to reset my chain , correct ? I just need to update geth

**Comment by karalabe** [comment]
Yup, updating the miner node is enough.

**Comment by sarbao** [comment]
Good morning all, I have just install 1.6.0 and things seem not to be working as they were. Amoung other issues ( cannot pre-fund ingenesis.json ) I cannot mine either. Geth crashed when I run miner.start(1); Any help?. Hello Karalabe back again :-)

**Comment by msolima** [comment]
@karalabe fixed! I just used the latest version. Did a git pull, now it's working.

@sarbao I'm using the latest commit in 1.6.0, I can mine, everything works fine. miner.start() doesn't crash. something might be wrong on your config.

**Comment by sarbao** [comment]
Good morning msolina,

  Yes something must be wrong but I am desperatly trying to find out. Could it be the genesis I am using ?. I would like to run like u 3 nodes 1 of them being the miner. But no succeed on the mining. I was able to connect all 3 nodes but when I start the miner nth is happening. I prefund the nodes succesfully but something as you said must be wrong cause I neved mined :-( ... its my 5th day with the issue aggg!. Appreciate if u can guide with this. Bets regards. SAlvador

**Comment by karalabe** [comment]
Make sure you didn't set a too high difficulty, causing you 1 core cpu
mining not to find any solutions in a reasonable time frame

**Comment by karalabe** [comment]
What difficulty did you set for your genesis block?

**Comment by sarbao** [comment]
Yes i put it at minimun 0x400 but still not working

**Comment by karalabe** [comment]
Windows? If yes, there was a memory mapping bug fixed before the 1.6.0
release. Please try eith the stable version. Also please share some logs,
preferrably running with --vmodule=ethash=5

**Comment by sarbao** [comment]
Yes it is windows.

Ok will do asap. Tks vm Peter and srry again for bothering so much

Bets regards

Salvador

**Comment by sarbao** [comment]
Hi Peter

Pls find enclosed the log file generated in the flow. WHen I run
miner.start(1); it started to generate DAG and after 10% crashed.

This is the step I took:

Step 1: geth --datadir "/tmp/dataDir/1" init
"/srv/tokenizer/genesis2GTH_161_reset.json"
Step 2: geth --datadir "/tmp/dataDir/2" init
"/srv/tokenizer/genesis2GTH_161_reset.json"

Step 3: geth --datadir "/tmp/dataDir/1" --networkid 12345001 --verbosity 3
 --vmodule=ethash=5 --ipcdisable --rpc --port 30301 --rpcport 8545
--rpcaddr localhost console 2>> /tmp/dataDir/nodo1.log
Step 4: geth --datadir "/tmp/dataDir/2" --networkid 12345001 --verbosity 3
 --vmodule=ethash=5 --ipcdisable --rpc --port 30302 --rpcport 8546
--rpcaddr localhost console 2>> /tmp/dataDir/nodo2.log

Step 5: I connect both nodes with admin.addPeer and they got connected
Step 6: I ran personal.newAccount on both
Step 7: I ran miner.start(1) and after a while it chrashed

I enclosed also my genesis

Bets reagrds

Salvador

**Comment by sarbao** [comment]
I put here the log file as I am not sure I sent over email

Regards
[dosNodos.zip](https://github.com/ethereum/go-ethereum/files/923641/dosNodos.zip)

**Comment by sarbao** [comment]
Srry by the way it is not only windows. The problem I am having know is in Ubuntu in AWS nodes. Srry misunderstanding

**Comment by sarbao** [comment]
Don't know if this could be a hint but when I run miner.gethashrate() i am
getting 0 ... still not mining :-(

**Comment by sarbao** [comment]
Agg nth, not able to mine. This ismy genesis.json
{
    "config": {
        "chainId": 15,
        "homesteadBlock": 0,
        "eip155Block": 0,
        "eip158Block": 0
    },
    "difficulty": "0x4000",
    "gasLimit": "0xffffffff",
    "nonce": "0x0000000000000042",
    "mixhash": "0x0000000000000000000000000000000000000000000000000000000000000000",
    "coinbase": "0x0000000000000000000000000000000000000000",
    "timestamp": "0x00",
    "parentHash": "0x0000000000000000000000000000000000000000000000000000000000000000",
    "extraData": "0x4000",
    "alloc": {}
}

And following all instructions. No errors on running geth simply miner.start(1); does nothing. THis is my geth command geth --datadir "/tmp/dataDir/1" --networkid 12345001 --verbosity 3  --vmodule=ethash=5 --ipcdisable --rpc --port 30301 --rpcport 8545 --rpcaddr localhost console 2>> /tmp/dataDir/nodo1.log 

I have both TCP and UDP enabled for 30300-30500 range ports but still not seeing peers if they are not on the same AWS instance and no mining by any mean. Don't know what I am doing wrong:

1.- Run geth with genesis.json
2.- RUn geth to create accounts
3.- Alloc the accounts in genesis
4.- Delete dataDir/numNode/geth and reload genesis to pre fund --> this is OK
5.- RUn the above command on same machine with diff port, rpcport and directory or in different mahcines
6.- If they are on same I run admin.addPeer I got true always but only connected on local
7.- None of the local instances nor different machines can mine

Don't know what I am doing wrongly in the flow. I have validated that flow in several private network and private blockchains  pages. Still trying !!!! 

Best regards and tks vm for help

Salvador
